src/sinks/pulsar.rs

```diff
@@ -0,0 +1,290 @@
+use crate::{
+    buffers::Acker,
+    event::{self, Event},
+    runtime::TaskExecutor,
+    sinks::util::MetadataFuture,
+    topology::config::{DataType, SinkConfig, SinkContext, SinkDescription},
+};
+use futures01::{
+    future, stream::FuturesUnordered, Async, AsyncSink, Future, Poll, Sink, StartSend, Stream,
+};
+use pulsar::{
+    proto::CommandSendReceipt, Authentication, Producer, ProducerOptions, Pulsar, SubType,
+};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::{collections::HashSet, path::PathBuf};
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("creating pulsar producer failed: {}", source))]
+    PulsarSinkFailed { source: pulsar::Error },
+    #[snafu(display("invalid path: {:?}", path))]
+    InvalidPath { path: PathBuf },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct PulsarSinkConfig {
+    address: String,
+    topic: String,
+    encoding: Encoding,
+    batch_size: Option<u32>,
+    #[serde(flatten)]
+    auth: Option<AuthConfig>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct AuthConfig {
+    auth_name: String,  // "token"
+    auth_token: String, // <jwt token>
+}
+
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Copy)]
+#[serde(rename_all = "snake_case")]
+pub enum Encoding {
+    Text,
+    Json,
+}
+
+pub struct PulsarSink {
+    topic: String,
+    encoding: Encoding,
+    producer: Producer,
+    pulsar: Pulsar,
+    in_flight: FuturesUnordered<MetadataFuture<SendFuture, usize>>,
+    // ack
+    seq_head: usize,
+    seq_tail: usize,
+    pending_acks: HashSet<usize>,
+    acker: Acker,
+}
+
+pub type SendFuture =
+    Box<dyn Future<Item = CommandSendReceipt, Error = pulsar::Error> + 'static + Send>;
+
+inventory::submit! {
+    SinkDescription::new_without_default::<PulsarSinkConfig>("pulsar")
+}
+
+#[typetag::serde(name = "pulsar")]
+impl SinkConfig for PulsarSinkConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        let sink = PulsarSink::new(self.clone(), cx.acker(), cx.exec())?;
+        let hc = healthcheck(self, &sink.pulsar)?;
+        Ok((Box::new(sink), hc))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "pulsar"
+    }
+}
+
+impl PulsarSink {
+    pub(crate) fn new(
+        config: PulsarSinkConfig,
+        acker: Acker,
+        exec: TaskExecutor,
+    ) -> crate::Result<Self> {
+        let auth = config.auth.map(|auth| Authentication {
+            name: auth.auth_name,
+            data: auth.auth_token.as_bytes().to_vec(),
+        });
+        let pulsar = Pulsar::new(config.address.parse()?, auth, exec).wait()?;
```

Comment:
Yes it could block. I didn't know of a better way to do this part. the tokio `0.1` executor provides `spawn` for types that return `()`. I'm not sure what else I could really do here given what is available. `PulsarSink::new` itself would have to run asynchronously for this to be made non-blocking. That would mean that `SinkConfig::new` would need to return a future also. Do you have any ideas?

Reply:
Unfortunately, no, just something we'll need to be aware of. Moving forward, we will likely convert these into `async fn` to be able to handle exactly this kind of issue but for now we will just eat the `wait`.